### PR TITLE
🌱 Annotate nodes with cluster info

### LINKED ISSUE #4044: Annotate nodes with the cluster name
**User Story**

As a user, I would like to know which CAPI cluster a node is part of by doing a quick inspection of its metadata (eg. `kubectl describe node`).

As a user, I would like to know if a k8s cluster was built using Cluster API.

**Detailed Description**

Add a `"cluster.x-k8s.io/cluster-name"` annotation on every node part of the cluster. This can be done in the machine and machine pool controllers.

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/assign

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: Adds basic cluster info to nodes as annotations to allow users to know if the node was created using cluster api and which cluster/namespace/machine or machine pool it belongs to.

Tested using release-0.3 branch:

```
 k --kubeconfig ./kubeconfig describe node capz-cluster-7-md-0-5f9fq
Name:               capz-cluster-7-md-0-5f9fq
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=Standard_D2s_v3
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=southcentralus
                    failure-domain.beta.kubernetes.io/zone=0
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=capz-cluster-7-md-0-5f9fq
                    kubernetes.io/os=linux
                    node.kubernetes.io/instance-type=Standard_D2s_v3
                    topology.kubernetes.io/region=southcentralus
                    topology.kubernetes.io/zone=0
Annotations:        cluster.x-k8s.io/cluster-name: capz-cluster-7
                    cluster.x-k8s.io/cluster-namespace: default
                    cluster.x-k8s.io/machine: capz-cluster-7-md-0-567986c6b8-xk8k8
                    cluster.x-k8s.io/owner-kind: MachineSet
                    cluster.x-k8s.io/owner-name: capz-cluster-7-md-0-567986c6b8
                    kubeadm.alpha.kubernetes.io/cri-socket: /run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    projectcalico.org/IPv4Address: 10.1.0.9/16
                    projectcalico.org/IPv4VXLANTunnelAddr: 192.168.194.128
                    volumes.kubernetes.io/controller-managed-attach-detach: true
``` 

```
 k --kubeconfig ./kubeconfig describe node capz-cluster-b-mp-0000001
Name:               capz-cluster-b-mp-0000001
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=Standard_D2s_v3
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=southcentralus
                    failure-domain.beta.kubernetes.io/zone=1
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=capz-cluster-b-mp-0000001
                    kubernetes.io/os=linux
                    node.kubernetes.io/instance-type=Standard_D2s_v3
                    topology.kubernetes.io/region=southcentralus
                    topology.kubernetes.io/zone=1
Annotations:        cluster.x-k8s.io/cluster-name: capz-cluster-b
                    cluster.x-k8s.io/cluster-namespace: default
                    cluster.x-k8s.io/owner-kind: MachinePool
                    cluster.x-k8s.io/owner-name: capz-cluster-b-mp-0
                    kubeadm.alpha.kubernetes.io/cri-socket: /run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    projectcalico.org/IPv4Address: 10.1.0.7/16
                    projectcalico.org/IPv4VXLANTunnelAddr: 192.168.219.0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
```                 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4044

### CODE REVIEW

File: controllers/machine_controller_noderef.go
Comment (vincepri) [inline controllers/machine_controller_noderef.go:95]: If we're not using our patch helper, we should probably check if there are any changes here before performing the patch, otherwise we'd end up patching at every reconciliation

**Comment by CecileRobertMichon** [comment]
/retest

MHC flake

File: util/annotations/helpers.go
Comment (vincepri) [inline util/annotations/helpers.go:60]: Could this be on a metav1 object instead?

File: exp/controllers/machinepool_controller_noderef.go
Comment (vincepri) [inline exp/controllers/machinepool_controller_noderef.go:114]: Same comment as above, re: patch helper

File: controllers/machine_controller_noderef.go
Comment (detiber) [inline controllers/machine_controller_noderef.go:92]: Would it also be helpful to annotate the Node with the kind/name of the owner if a controller reference exists on the Machine here?

That would help link the Node more directly to a MachineSet or KubeadmControlPlane (if the Machine is managed by either of those).

Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go]: Added

File: util/annotations/helpers.go
Comment (CecileRobertMichon) [reply util/annotations/helpers.go]: indeed, changed

File: controllers/machine_controller_noderef.go
Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go]: done

**Comment by CecileRobertMichon** [comment]
Thanks for the review @vincepri @detiber. The change ended up being a bit bigger than I expected because I added some logic to fetch the Machine owner but I think it's better now. Added unit test coverage for the new helper/utils functions.

File: util/util.go
Comment (CecileRobertMichon) [inline util/util.go:348]: /hold

I need to fix this to use the controlplane Group. Unfortunately, there's an import cycle between controlplane/api and util. I will fix the cycle in a separate PR to unblock this.

File: exp/controllers/machinepool_controller_noderef.go
Comment (detiber) [inline exp/controllers/machinepool_controller_noderef.go:117]: Similar to above wrt `MergeFrom`

File: controllers/machine_controller_noderef.go
Comment (detiber) [inline controllers/machine_controller_noderef.go:98]: I don't think we actually need to go this far, I think we could do something like this instead:

```suggestion
	if owner := metav1.GetControllerOfNoCopy(obj); owner != nil {
		desired[clusterv1.OwnerKindAnnotation] = owner.Kind
		desired[clusterv1.OwnerNameAnnotation] = owner.Name
	}
```

File: util/util.go
Comment (detiber) [reply util/util.go:348]: I'm not sure there is a need to enforce certain allowed GVKs here, any controller reference should be sufficient for our purpose, since it would indicate ownership of the lifecycle of the resource (that it wasn't created by a user).

I do wonder if instead of just adding the Kind as the value of the annotation if it would make more sense to use GVK instead to avoid any potential for name conflicts/confusion.

File: api/v1alpha4/common_types.go
Comment (fabriziopandini) [inline api/v1alpha4/common_types.go:44]: I think that in case of MachinePools we should not add a different annotation but instead add something like

```
cluster.x-k8s.io/owner-kind: MachinePool
cluster.x-k8s.io/owner-name: capz-cluster-2-mp-0
```

So all the nodes managed by higher-level abstraction have a consistent set of labels.
The only exception for the machine pools is the `cluster.x-k8s.io/machine` being missing

File: controllers/machine_controller_noderef.go
Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go:98]: True, this works and is a lot simpler 😅 thanks!

File: api/v1alpha4/common_types.go
Comment (vincepri) [reply api/v1alpha4/common_types.go:44]: Given that these are annotations, could we instead json-marshal the owner reference as we see it today instead of creating a new standard to store this information?

File: controllers/machine_controller_noderef.go
Comment (vincepri) [reply controllers/machine_controller_noderef.go:100]: Could we potentially also make use of the patch helper here?

Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go:100]: @vincepri in this case what value does patch helper bring that can't be accomplished by MergeFrom?

Comment (vincepri) [reply controllers/machine_controller_noderef.go:100]: Patch helper checks if there are any changes before issuing the patch, IIRC merge from won't unless you explicitly check

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [reply api/v1alpha4/common_types.go:44]: what do you mean by json-marshal the owner reference? How would it be represented in the annotations? Would it be one annotation for the entire owner ref?

Comment (vincepri) [reply api/v1alpha4/common_types.go:44]: Something like `cluster.x-k8s.io/owner` and its value would be the json value of an OwnerReference object

File: util/util.go
Comment (CecileRobertMichon) [reply util/util.go]: removed this func altogether

File: controllers/machine_controller_noderef.go
Comment (CecileRobertMichon) [reply controllers/machine_controller_noderef.go]: switched to patch helper

File: api/v1alpha4/common_types.go
Comment (CecileRobertMichon) [reply api/v1alpha4/common_types.go]: IMO 

```
cluster.x-k8s.io/owner-kind: MachinePool
cluster.x-k8s.io/owner-name: capz-cluster-2-mp-0
``` 

is more readable than dumping the json string in there. I don't really care either way, what I really wanted from this PR in the first place was the cluster name, owner name + kind were just added on for additional info.

Comment (vincepri) [reply api/v1alpha4/common_types.go]: Sounds good to me, I assume that would depend on who's using these and how, for example if this is only for human users that's probably fine. For machines / controllers the information above for example is missing the API group (this is only a problem if we ever say that a Machine could be controlled outside of Cluster API main types)

File: util/util.go
Comment (CecileRobertMichon) [reply util/util.go]: /hold cancel

**Review by vincepri — APPROVED** [review]
/lgtm
/milestone v0.4.0

**Pull request merged**
